shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,313 @@
+         *
+         * @tparam DSL the NUClear dsl for the on statement
+         *
+         * @param task The reaction task object that just finished
+         */
+        template <typename DSL>
+        static inline void postcondition(NUClear::threading::ReactionTask& task) {
+            // Take the task id and send it to the Director to let it know that this Provider is done
+            task.parent.reactor.emit(std::make_unique<commands::ProviderDone>(task.parent.id, task.id));
+        }
+    };
+
+    /**
+     * This DSL word used to define an entry transition into a Provider.
+     * It should normally be coupled with a Causing DSL word.
+     * When coupled with a Causing DSL word and a Provider that has a When condition, this entry will run in place of
+     * the Provider until the When condition is fulfilled.
+     * If it is used without a Causing DSL word it will run once as a Provider before the actual Provider is executed.
+     *
+     * @tparam T the Provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Entering : public ProviderBase<T, commands::ProviderClassification::ENTERING> {};
+
+    /**
+     * This DSL word is used to define a leaving transition from a Provider.
+     * It should normally be combined with a Causing DSL word.
+     * In this case when another Provider needs control and uses a When condition, this will run until that
+     * condition is fulfilled. A Leaving condition will be preferred over an Entering condition that provides the
+     * same Causing statement (the system will rather a module cleans up after itself than ask another module to do
+     * so).
+     *
+     * @tparam T the Provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Leaving : public ProviderBase<T, commands::ProviderClassification::LEAVING> {};
+
+    /**
+     * This DSL word is used to define a Provider for a type.
+     * It will execute in the same way as an on<Trigger<T>> statement except that when it executes will be
+     * determined by the Director. This ensures that it will only run when it has permission to run and will
+     * transition between Providers in a sensible way.
+     *
+     * @tparam T the Provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Provides : public ProviderBase<T, commands::ProviderClassification::NORMAL> {};
+
+    /**
+     * The When DSL word is used to limit access to a Provider unless a condition is true.
+     * This will prevent a Provider from running if the state is false.
+     * However if there is a Entering<T> or Leaving reaction that has a Causing relationship for this When, it will
+     * be executed to ensure that the Provider can run.
+     *
+     * @tparam T the condition expression that must be true in order to execute this task.
```

Comment:
This does not line up with the actual template paramters

Reply:
Indeed